File: 백준/Gold/2493. 탑/탑.cc
// stack

#include <iostream>
#include <vector>
#include <stack>

using namespace std;
using pii = pair<int,int>;

int n;
int arr[500001];
int ans[500001];
stack<pii> st;

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);

    cin >> n;
    for (int i = 1; i <= n; i++) cin >> arr[i];

    for (int i = n; i > 0; i--) {
        st.push({arr[i], i});
        while (!st.empty()) {
            if (arr[i-1] > st.top().first) {
                ans[st.top().second] = i - 1;
                st.pop();
            } else break;
        }
    }

    for (int i = 1; i <= n; i++) cout << ans[i] << " ";

    return 0;
}

File: 백준/Gold/2493. 탑/탑.py
import sys

n = sys.stdin.readline()
arr = sys.stdin.readline().split()
st = []
ans = []
n = int(n)
# int 변환
for i in range(n):
    arr[i] = int(arr[i])
    ans.append(0)

for i in range(n-1, -1, -1):
    # 현재 스택에 나보다 작은 거 전부 pop
    while len(st) != 0 and arr[st[-1]] < arr[i]:
        ans[st[-1]] = i + 1
        st.pop()
    st.append(i)

for i in ans:
    print(i, end=' ')